File: Implementation/service_lane.java
import java.util.*;

class service_lane {
   private static List<Integer> widths = new ArrayList<>();

   public static void main(String[] args) {
      process(read());
   }

   private static List<Case> read() {
      List<Case> cases = new ArrayList<>();
      Scanner scanner = new Scanner(System.in);
      int freewayLength = scanner.nextInt();
      int caseCount = scanner.nextInt();
      for (int i = 0; i < freewayLength; i++)
         widths.add(scanner.nextInt());
      for (int i = 0; i < caseCount; i++)
         cases.add(new Case(scanner));
      scanner.close();
      return cases;
   }

   private static void process(List<Case> cases) {
      for (Case cas : cases) {
         // System.out.println(cas);
         System.out.println(cas.process());
      }
   }

   // class is static so objects can be instantiated in read() a static method
   static class Case {
      int entryIndex, exitIndex;

      public Case(Scanner scanner) {
         entryIndex = scanner.nextInt();
         exitIndex = scanner.nextInt();
      }

      public int process()
      {
         // widths[entryIndex..exitIndex] is the range in which to find the minimum
         List<Integer> subWidths = widths.subList(entryIndex, exitIndex+1);
         // make a copy of the range
         List<Integer> copy = new ArrayList<Integer>(subWidths);
         // sort the copy
         Collections.sort(copy);
         // return the first entry of the sorted copy
         return copy.get(0);
      }

      @Override
      public String toString() {
         return entryIndex + " " + exitIndex;
      }
   }
}

File: Implementation/service_lane.rb
def process(widths, cases)
  minimums = []
  # test[0] is entry; test[1] is exit. widths[entry..exit] is the range in
  # which to find the minimum. all is needed is to sort the sub-array
  # widths[entry..exit], then take the first entry of the sorted sub-array
  cases.each do |test|
    minimums << widths[test[0]..test[1]].sort.first
  end
  return minimums
end

counts = gets.split.map {|i| i.to_i}
widths = gets.split.map {|i| i.to_i}
cases = []
counts[1].times do
  cases << gets.split.map {|i| i.to_i}
end

process(widths, cases).each do |minimum|
  puts(minimum)
end
